Add observers for Poisson system

nilsvu opened the pull request:
### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [x] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [x] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [x] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu committed:
Add observers for Poisson system

mar-celine commented on src/Elliptic/Systems/Poisson/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+#include "Utilities/TaggedTuple.hpp"
+
+namespace Poisson {
+namespace Actions {
```
For comparison, ScalarWave appears in the namespaces documentation with Actions and Solutions. I only see Solutions for Poisson, should Actions be added as well?

nilsvu replied:
It should be, but not in this file. Same with `Evolution/GrMhd/ValenciaDivClean`. The fixup commit below proposes to move this into an `Actions` subdirectory with a file `Actions.hpp` that documents the namespace. Is this an approach we want to follow?

mar-celine commented on src/Elliptic/Systems/Poisson/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+    DataVector field_error = get(field) - get(field_analytic);
+
+    // Retrieve linear residual
+    // We have to reduce it here again, even though it should be the same on all
```
This is probably because I am unfamiliar with the code, but what does "it" refer to in this comment? Also, is the residual itself linear or could "linear residual" be better phrased as "The residual from the linear solver"?

nilsvu replied:
rephrased it

mar-celine commented on src/Elliptic/Systems/Poisson/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+        db::add_tag_prefix<LinearSolver::Tags::Residual, fields_tag>>;
+    const double linear_residual = get<residual_magnitude_tag>(box);
+
+    // Compute l2 error square over local element
```
I would go with "squared" over "square" if that is what we have been doing when documenting squared errors

mar-celine commented on src/Elliptic/Systems/Poisson/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+
+    // Compute l2 error square over local element
+    const double local_l2_error_square =
+        definite_integral(square(field_error), mesh);
```
I have noticed that we compute L2 norms differently in other Observers, but this method of computing the L2 norm does not seem unreasonable to me, so I don't have an immediate objection. However, I think we could gain from carefully revisiting how L2 norms are computed numerically, ideally with a source from the literature if someone knows offhand where it is described, so that we can all be on the same page and provide future people with a reference if they have a question about this is well.

nilsvu replied:
I think so too and actually opened #1189 for this a while back.

mar-celine commented on src/Elliptic/Systems/Poisson/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+                            get(field_analytic));
+    components.emplace_back(element_name + Poisson::Field::name() + "Error",
+                            field_error);
+    components.emplace_back(element_name + "InertialCoordinates_x",
```
Should this be either of "inertial_coordinates_x" or "InertialCoordinatesX"? I'm unsure if we have a style convention for names in strings, but since file names are CamelCased I would also CamelCase this.

nilsvu replied:
This naming convention is used in the other observers. I think the idea is to have the label in CamelCase and append the tensor component (not part of the label) with an underscore.

mar-celine replied:
Thank you for the clarification :+1: I will keep this in mind along with the other naming conventions listed in the code review guidelines.

nilsvu commented:
@gsb76 could you check the fixups and see if there is anything else?

moxcodes commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+namespace Actions {
+
+/*!
+ * \brief Temporary action to observe volume and reduction data for the Poisson
```
what is `Temporary action`?

nilsvu replied:
see below

moxcodes commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+            .field_variables(inertial_coordinates));
+
+    // Compute error between numeric and analytic solutions
+    DataVector field_error = get(field) - get(field_analytic);
```
this can probably be `const`

moxcodes commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+    const double linear_residual = get<residual_magnitude_tag>(box);
+
+    // Compute l2 error squared over local element
+    const double local_l2_error_square =
```
This will give an (un-normalized) integral over the logical coordinates. So, if `field_error` is 1, this will give 2^dim. Should this be normalized? if not, we'd prefer a comment explaining what this is intended to be.

nilsvu replied:
changed to the vector L2 norm instead of the integral for now. we can discuss this in #1189 if there's more to it

moxcodes commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+    components.emplace_back(element_name + Poisson::Field::name() + "Error",
+                            field_error);
+    components.emplace_back(element_name + "InertialCoordinates_x",
+                            inertial_coordinates.get(0));
```
this can be a compile-time `get<0>` (and below)

moxcodes commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,124 @@
+    using observed_reduction_data = Parallel::ReductionData<
+        Parallel::ReductionDatum<size_t, funcl::AssertEqual<>>,
+        Parallel::ReductionDatum<size_t, funcl::Plus<>>,
+        Parallel::ReductionDatum<double, funcl::AssertEqual<>>,
```
Asserting that `double`s are equal can be spooky - doubles can end up slightly different on different processors, if computed independently, etc. Are we sure this will be safely identical across cores (e.g. via a singleton)? If yes, can this be commented on? If not, the code should probably be changed.

nilsvu replied:
this is indeed identical and computed in a singleton, see comment in the code. Will be sent to the observer directly from the singleton when #1198 is resolved.

nilsvu replied:
removed this, since #1233 is merged

moxcodes reviewed (changes requested):
Also, we'd like a test for the observer. See, for instance, `tests/Unit/IO/Observers/Test_ReductionObserver.cpp` for an existing observer test.
Group code review with @markscheel @fmahebert @roboberlin @geoffrey4444

nilsvu commented:
@markscheel @fmahebert @roboberlin @geoffrey4444 This is a "temporary action" in the sense that people may want to change it before compiling, just like they would change the analytic solution/data in the Metavariables. In the action you can specify what data should be written to disk. Therefore we can write no useful unit tests for it, just like we can't for the Metavariables. It is the same as `ValenciaDivClean/Observe.hpp`, which is also not tested.
Eventually I think we should reduce the need for compile-time changes, but this PR is not the place to discuss this, perhaps look at #1219.

mar-celine commented:
@nilsleiffischer feel free to squash in the changes related to my comments :octocat:

wthrowe commented:
Much of the ValenciaDivClean stuff is untested hacks because that part of the code was under deadlines for the accretion disk paper.  Don't use anything in there as an example of good practice.

nilsvu commented:
@wthrowe I understand. #1207 (and less so also #1205) try to improve on this for the elliptic code, e.g. adding tests for the initialisers. However, this observe action we can't really test for the reasons I give above. Do you agree, or do you have objections to that?

wthrowe commented:
I don't see why those things prevent you from writing a test.  We do test the metavariables as part of the InputFiles.*.parse tests, and those tests do fail if you modify the metavariables.

Also, "temporary" doesn't seem like the right word for what you are describing, although I'm not coming up with a better one off the top of my head.

nilsvu commented:
@wthrowe You're right, of course. I tried to talk my way out of having to write this test but failed 😄 So here it is

nilsdeppe commented:
@wthrowe @moxcodes @gsb76 could you please prioritize looking at the fixups for this PR?

moxcodes commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,142 @@
+                            get<0>(inertial_coordinates));
+    if (Dim >= 2) {
+      components.emplace_back(element_name + "InertialCoordinates_y",
+                              inertial_coordinates.get(1));
```
it seems like these could also be compile-time `get<1>` (and below for 3-dimensional case)

mar-celine replied:
For my own understanding, do compile-time `get`s work inside of an `if` statement conditional on the `Dim`?

moxcodes replied:
oops, good point @gsb76 . This would only work if we had constexpr if's. My suggestion won't work without fancier tricks that won't be worth it for this code segment (`make_overloader` would do it, or if we could use the c++17 language features).

moxcodes commented on tests/Unit/Elliptic/Systems/Poisson/Actions/Test_Observe.cpp:
```diff
@@ -0,0 +1,302 @@
+  for (const auto& id : element_ids) {
+    tuples::get<ElementMockDistributedObjectsTag>(dist_objects)
+        .emplace(ElementIndex<2>{id},
+
```
possibly unintentional newline in the middle of the emplace?

moxcodes commented on tests/Unit/Elliptic/Systems/Poisson/Actions/Test_Observe.cpp:
```diff
@@ -0,0 +1,302 @@
+    runner.invoke_queued_simple_action<obs_component>(0);
+  }
+
+  const std::string reduction_h5_file_name = reduction_file_name + ".h5";
```
is it possible that these file checks can be moved to just before the `obs_writer` invocations and merge the for loops?

nilsvu replied:
No I don't think the for loops can be merged. I assume the registration needs to be done before the elements start contributing. @nilsdeppe is that correct?
In any case, I don't think that change would actually improve this test, since it does not reflect the order in which the actions would be invoked in the executables and isn't important for performance.

moxcodes replied:
That's fine, then. I'd prefer the test to be as similar in structure to its eventual use case as possible, so given that this is the intended pattern for calling code, it's better left as is.

nilsdeppe replied:
Yes, registration must complete before anything is sent.

moxcodes commented on tests/Unit/Elliptic/Systems/Poisson/Actions/Test_Observe.cpp:
```diff
@@ -0,0 +1,302 @@
+      db::compute_databox_type<tmpl::append<simple_tags, compute_tags>>;
+};
+
+struct AnalyticSolution {
```
it seems undesirable in terms of testing robustness to have the `AnalyticSolution` you're writing be identical to the also written x-coordinate value. You could imagine a bug where the incorrect thing is written to or retrieved from the file that this wouldn't disambiguate. A simple fix might be to just perform some simple mathematical operation here to make sure that you get different values in each slot, and update the test below.

nilsvu replied:
you're right, fixed it

moxcodes reviewed (dismissed):
A few minor nitpicks and suggestions. In light of new code review guidelines, I won't hold things up for these small details. I'm happy squashing the fixups and any responses to my comments.

mar-celine commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,142 @@
+/*!
+ * \brief Action to observe volume and reduction data for the Poisson system
+ *
+ * This actions observes the following:
```
This actions -> This action

mar-celine commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,142 @@
+ *     u^\mathrm{numeric}_i - u^\mathrm{analytic}_i\right)^2}\f$ over the grid
+ *     points across all elements.
+ * - Volume data:
+ *   - `Poisson::Field::name()`: The numeric solution \f$u^\mathrm{numeric}\f$
```
I would change numeric to numerical because I think typically we've been describing solutions as either analytic or numerical (as opposed to numeric)

mar-celine commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,142 @@
+ *   - `Poisson::Field`
+ *   - `Tags::Coordinates<Dim, Frame::Inertial>`
+ *
+ * \note This action can be adjusted before compiling an executable to observe
```
If we still have the behavior where the observed quantities must match in the Metavariables as well as in the Observe Action, should this say "This Action must be adjusted to match the Metavariables before compiling to observe only the desired quantities?" 

Edit: This is more of a question for myself, feel free to not change your `\note` if you feel it is accurate

nilsvu replied:
That the `Metavariables::reduction_data_tags` must reflect this is not desired behaviour and will be remedied at some point. I'd prefer not to add that comment here.

mar-celine commented on tests/Unit/Elliptic/Systems/Poisson/Actions/Test_Observe.cpp:
```diff
@@ -0,0 +1,302 @@
+namespace {
+template <typename Metavariables>
+struct MockElementArray {
+  using component_being_mocked = void;  // Not needed
```
I would change it to "There is no component being mocked, but the type alias is still needed, so it is set to void" or something similar that is correct if what I suggested is incorrect - "Not needed" on its own makes me think "Then why is it here?" Or more concisely: `// No component is being mocked`

nilsvu replied:
I copied this from elsewhere, and don't have enough insight into when this alias is actually needed to add a more useful comment here. I know there are plans to improve the documentation for the action testing framework though.

mar-celine commented:
Feel free to squash the current fixups as well as the fixes to my comments :smile_cat: Removing the `updated` tag for now since both  reviewers of have made their comments on the fixups.

nilsvu commented:
@gsb76 @moxcodes addressed your comments and left the last fixup. could you quickly confirm to squash that? (I know you already said so @gsb76, but @moxcodes hasn't)

moxcodes commented:
Yep, I'm happy. Go ahead and squash.

nilsvu commented:
👍 this should be done then @sxs-collaboration/spectre-core-devs

nilsvu commented on tests/Unit/Elliptic/Systems/Poisson/Actions/Test_Observe.cpp:
```diff
@@ -0,0 +1,301 @@
+    // NumberOfPoints
+    CHECK(reduction_data(0, 1) == 18);
+    // L2Error
+    CHECK(reduction_data(0, 2) == sqrt(24.));
```
This should have been `sqrt(72.)` (3 elements * 6 points * 2^2 = 72., forgot the 3 elements before). Fixed and squashed.

nilsdeppe commented on src/Elliptic/Systems/Poisson/Actions/Observe.hpp:
```diff
@@ -0,0 +1,143 @@
+                            get(field_analytic));
+    components.emplace_back(element_name + Poisson::Field::name() + "Error",
+                            field_error);
+    components.emplace_back(element_name + "InertialCoordinates_x",
```
The GRMHD executables handles the different coordinates as:
```cpp
      for (size_t d = 0; d < Dim; ++d) {
        const std::string component_suffix =
            d == 0 ? "_x" : d == 1 ? "_y" : "_z";
```
I'm not requesting a change, just offering this up. Please reply with whether or not you're going to leave as is or change it :)

nilsvu replied:
I liked the `if`s better since it's more clear what it does. No strong opinion but I prefer to leave it as is.

nilsdeppe reviewed (dismissed):
I have one comment that offers a suggestion that you may or may not care about. Once you decide what you want to do there I'm fine with this moving forward whatever you decide :)

nilsdeppe commented:
@nilsleiffischer  IWYU is unhappy

mar-celine reviewed (dismissed):
Stray typo on line 37 of Observe.hpp, (`anlytic`) feel free to fix it if you need to recommit/push anyways, otherwise I can do it afterwards :+1:

wthrowe commented:
Looks fine, aside from the IWYU problem mentioned above.

nilsvu commented:
done @sxs-collaboration/spectre-core-devs. Restarted the segfaulted Travis builds.
